File: LeetCode_20/Solution_ZhongJing.kt
package LeetCode_20

import java.util.LinkedList

/**
 * 日期：2023-07-02
 * 作者：仲景
 */
class Solution {
    fun isValid(s: String): Boolean {
        val m = mapOf<Char, Char>(
            ')' to '(',
            ']' to '[',
            '}' to '{',
        )

        val stack = LinkedList<Char>()
        s.forEach {
            if (it == '(' || it == '[' || it == '{') {
                stack.push(it)
            } else {
                if (stack.isEmpty() || stack.pop() != m[it]) {
                    return false
                }
            }
        }

        return stack.isEmpty()
    }
}

File: LeetCode_20/Solution_ZhongJing.py
class Solution:
    """
    日期：2023-08-04
    作者：仲景
    """
    
    def isValid(self, s: str) -> bool:
        # map
        sMap = {
            ")": "(",
            "]": "[",
            "}": "{",
        }
        # 栈
        stack = []
        # 遍历s
        for c in s:
            # 如果是左括号，入栈
            if c in ["(", "[", "{"]:
                stack.append(c)
                # 如果是右括号，出栈，判断
            else:
                if len(stack) == 0 or stack.pop() != sMap[c]:
                    return False
        return len(stack) == 0